### ISSUE TITLE
Add capability for recursive least squares (RLS)

### ISSUE THREAD

**Issue by gonike** [scipy/scipy#8682]
Is there any particular reason why there is no capability for linear recursive least squares (RLS) in scipy? It is an absolute standard signal processing / data analysis algorithm, and I was somewhat surprised to find that it isn't implemented in scipy. Or am I completely missing something? Of course implementing RLS in python is relatively trivial, but I still think it would be nice to have a well-tested, performant implementation available in scipy.

#### RLS vs. standard LS 
There are three reasons apparent to me why someone could be interested in RLS as opposed to standard LS (i.e. `scipy.linalg.lstsq`):
1. Memory issues. When dealing with very large (dense) datasets, matrices can become huge and memory issues may arise when using standard LS. This is of no concern to RLS, since it does not work with the full data matrix but rather using data slices.
2. Estimating time-varying parameters. RLS is easily supplemented by a forgetting factor that allows for time-varying parameter estimates that are very useful in many applications.
2. Prototyping for embedded devices. As said above, RLS is an absolute standard for (real-time) signal processing applications. It would be very useful to have RLS (with an optional forgetting factor) available in scipy to be able to prototype and test algorithms.

#### Implementation
I have to say that I am by no means an expert regarding efficient numerical implementation, hence I am completely unsure which of the following variants would be desirable for implementation - maybe both?
* Naive implementation of the basic RLS algorithm (https://en.wikipedia.org/wiki/Recursive_least_squares_filter#RLS_algorithm_summary), due to, e.g., Ljung (1999): System Identification. 
* Implementation of more efficient and numerically stable versions, e.g. as exposed in the [The Digital Signal Processing Handbook](https://www.crcpress.com/The-Digital-Signal-Processing-Handbook-Second-Edition---3-Volume-Set/Madisetti/p/book/9781420045635). Also see [the wiki entry on lattice RLS filters](https://en.wikipedia.org/wiki/Recursive_least_squares_filter#Lattice_recursive_least_squares_filter_(LRLS)).

Any comments?

**Comment by josef-pkt** [comment]
as background:
statsmodels has an implementation based on kalman filter statespace model
http://www.statsmodels.org/dev/generated/statsmodels.regression.recursive_ls.RecursiveLS.html
this was supposed to be a replacement for pandas expanding OLS.

I don't remember if we have anything with time varying (forgetting) coefficients that is a ready-made model. The statespace framework in statsmodels is pretty good and is extensible, but only handles dense arrays.

Nevertheless, there might be applications in signal where a custom made RLS or similar would be more appropriate, (scipy now also has some linalg up and down dating functions that might be useful for this.)

As example, a long time ago (before statsmodels had the statespace framework), I wrote a version that was mainly used for structural change, CUSUM, tests.

**Comment by rgommers** [comment]
The question has never been asked before as far as I can remember; no one has proposed to add it to SciPy. Despite the fairly general name, RLS does seem quite specific to signal processing. In terms of adaptive filtering `scipy.signal` also doesn't have much, so not surprising it doesn't have RLS either.

I'm not an expert, but a good well-tested algorithm is probably welcome as addition to `signal`.

**Comment by gonike** [comment]
Interesting, so how would I go about pursuing this further? Providing a Python implementation of one or two standard RLS algorithms and some tests shouldn't be too big a problem, but is that what's required? (I couldn't find anything related in LAPACK or netlib.) Should I write to the mailing list first?

**Comment by ilayn** [comment]
There are some implementations available outside SciPy, say, https://github.com/matousc89/padasip

**Comment by gonike** [comment]
Good point @ilayn. I had seen that before but just gave it a closer look. This does look good, and the implementations of RLS and (N)LMS are probably reasonably close to what one would like to do here. Would including this in SciPy be an option? License is MIT.

@rgommers I had seen the statsmodels RLS, but although I'm not 100% sure what exactly they're doing internally, I think their (state-space based) implementation differs strongly from standard RLS.

**Comment by ilayn** [comment]
To my very opinionated eyes adaptive filters are a bit too specific for scipy.signal but I'd better leave that decision to scipy.signal experts.

**Comment by holy-motors** [comment]
I'm no expert at all but to summarise the discussion, 
1. Everyone agree that RLS should be included in scipy

And we have two options for the implementation
a) Include the implementation from Padasip to Scipy as @gonike [suggested](https://github.com/scipy/scipy/issues/8682#issuecomment-378908571)

b) A new implementation, possibly in Cython  as @josef-pkt [suggested ](https://github.com/scipy/scipy/issues/8682#issuecomment-378917061)

**Comment by cannontwo** [comment]
I have a bare-bones implementation of recursive least-squares in NumPy available at https://github.com/cannontwo/rls. Is this an open area for contribution? I could go about putting together a pull request if so.

**Comment by rgommers** [comment]
Thanks for asking @cannontwo. There is clearly interest, so it would be nice to see a good implementation in `scipy.signal`. What we need I think is a comparison between all the available implementations:
- `statsmodels`
- `padasip`
- another one I just found in https://github.com/LCAV/pyroomacoustics/blob/pypi-release/pyroomacoustics/adaptive/rls.py
- your own

If we are going to include RLS in SciPy, it should be the most performant one (and stable/correct of course).

> However, to be useful in batch processing, I think RecursiveLS needs to be in Cython. (The Kalman filter in statsmodels was too slow without using cython with direct access to LAPACK through the scipy cython wrappers.)

After comparing the existing implementations, some benchmarking of the fastest one may shed some light on this.
